src/Pos/Wallet/Web/Server/Methods.hs

```diff
@@ -1034,6 +1042,45 @@ testResetAll | isDevelopment = deleteAllKeys >> testReset
         keyNum <- length <$> getSecretKeys
         replicateM_ keyNum $ deleteSecretKey 0
 
+----------------------------------------------------------------------------
+-- JSON backup methods
+----------------------------------------------------------------------------
+
+restoreWalletFromBackup :: WalletWebMode m => WalletBackup -> m CWallet
+restoreWalletFromBackup WalletBackup {..} = do
+    let wId = addressToCId . makePubKeyAddress $ encToPublic wbSecretKey
+        (WalletMetaBackup wMeta) = wbMeta
+        accList = HM.toList wbAccounts
+                  & each . _2 %~ \(AccountMetaBackup am) -> am
+
+    addSecretKey wbSecretKey
+    for_ accList $ \(idx, meta) -> do
+        let aIdx = fromInteger $ fromIntegral idx
+            seedGen = DeterminedSeed aIdx
+        accId <- genUniqueAccountId seedGen wId
+        createAccount accId meta
+
+    selectAccountsFromUtxoLock @WalletSscType [wbSecretKey]
+    createWalletSafe wId wMeta
```

Comment:
I did it this way to return wallet with correct balance without needing to get it again after all. Accounts, afaiu, does not depend on presence of wallet metadata in wallet-db, so it doesn't seem unsafe to me in some sense (maybe I'm missing something, some concurrency issues?). Anyway, it worked when I tested it.